CONTRIBUTING.md

```diff
@@ -66,6 +66,111 @@ For first-time contributers, we suggest to look for issues categorized as [trivi
 
 If you want a hassle-free contribution look for issues categorized as [preapproved](https://issues.dlang.org/buglist.cgi?component=dmd&keywords=preapproved&product=D).
 
+## DMD Style
+
+Here is a short list of stylistic issues the core team will expect in
+pull requests. Much of the source code does not follow these, but
+we expect new code to, and PRs retrofitting the existing code to
+follow it is welcome.
+
+1. Use attributes `const`/`nothrow`/`pure`/`scope`/`@safe`/`private`/etc.
+Successfully using `pure` functions is regarded with particular favor.
+
+2. Use correct Ddoc function comment blocks. Do not use Ddoc comments for
+overrides unless the overriding function does something different (as far as
+the caller is concerned) than the overridden function. Ddoc comment blocks
+are often overkill for nested functions and function literals; use ordinary
+comments for those.
+
+3. Do not use `strlen`/`strcmp` and their ilk. Use D arrays instead.
+
+4. Use `ref` instead of raw pointers.
+
+5. Use nested functions to get rid of rats' nests of goto's.
+
+6. Look for duplicative code and factor out into functions.
+
+7. Declare local variables as `const` as much as possible.
+
+8. Use Single Assignment for local variables:
+```
+T t = x;
+...
+t = y;
+ ...
+```
+becomes:
+```
+T tx = x;
+...
+T ty = y;
+...
+```
+
+9. "Shrinkwrap" the scope of local variables as tightly as possible
+around their uses.
+
+10. Similar to (8), use distinct variable names for non-overlapping uses.
+
+11. Avoid use of mutable globals as much as practical. Consider passing them
+in as parameters.
+
+12. Avoid use of default parameters. Spell them out.
+
+13. Minimize use of overloading.
+
+14. Avoid clever code. Anybody can write clever code. It takes genius to write
+simple code.
+
+15. Try to reduce cyclomatic complexity, i.e. think about how to make the code work
+without control flow statements.
+
+16. Try not to mix functions that "answer a question" with functions that
+"mutate the data".
+This was done successfully in src/dmd/escape.d, it wasn't easy, but
+it was well worth it.
+
+17. Try to eliminate reliance on `global.errors`.
+
+18. For aggregates that expose public access to fields, think hard about why this is
+necessary and if it can be done better. Merely replacing them with read/write properties
+accomplishes nothing. The more of its internal details can be made private, the better.
+
+19. Try to use function prefixes:
+
+ * `is` is the parameter in a certain category?
+ * `has` does the parameter have a certain feature?
+ * `can` for can I do X with the parameter?
+
+Such functions should not be mutating the data nor issuing error messages.
+
+20. The function return value variable should be named `result`.
+
+21. The more constrained the scope of a name is, the shorter it should be.
+
+
+
+## The following will not be viewed with favor:
+
+1. Grand renamings of variables and functions
+
+2. Shuffling all the code about
+
+3. Creating packages
```

Comment:
Ditto. I'd love to reorganise the compiler into ast, sema ...

Reply:
That's in the negative column :-)

Reply:
I know, hence "ditto (I don't think this is a good idea.)" . Creating packages is something we should be doing not something we should be avoiding. 

`src/dmd/` contains ~120 d source files in one directory, with no organisation. This is getting unwieldy. Taking a curry glance I see at least the following packages:

* AST (e.g. `aggregate.d`)
* backend interfacing (`e2ir.d`, `s2ir.d`, `toir.d`) (incidentally the are removed from LDC's copy of the front end)
* optimisations (`constfold.d`, `inline[|cost].d`) 
* iasm
* lib (`lib[|elf|mach|mscoff|omf].d`) (so are these)
* visitors
* scan (`scan[elf|mach|mscoff|omf].d`) (so are these)
* semantic analysis (`semantic[2|3].d`, `typesem.d`, `statementsem.d`, `templateparamsem.d`
* utils (`util.d`, `filecache.d`, `intrange.d`)

We should also get rid of the headers into their own folder (I mean we should get rid of them all together when we can generate them, but we're not there yet).

I agree that we shouldn't over-package the codebase, but we have the opposite problem at the moment. Compare [clang](https://github.com/llvm-mirror/clang/tree/master/lib) with [dmd](https://github.com/dlang/dmd/tree/master/src/dmd)

Reply:
Been there before: https://github.com/dlang/dmd/pull/7639 :/
I don't know anyone who's happy with the status quo (except for Walter, of course).

Reply:
Well I've put this on the AGM agenda.

Reply:
The reason I don't like packages for dmd is because of `grep`.

Reply:
`package` and `module` are D's units of encapsulation.  They should be embraced and leveraged (judiciously and with good taste, of course).  Why would organizing things in packages negatively affect utilizing `grep`?

Reply:
How long have we been using git for now? https://git-scm.com/docs/git-grep

Functions can and will have the same name regardless of whether they are in a module or a package.

Reply:
> The reason I don't like packages for dmd is because of grep

How is this a problem?
`grep` does have a `-r` flag for recursive searching and everyone is using `ack` or `ag` these days anyhow, e.g.

https://github.com/ggreer/the_silver_searcher

Reply:
It's not impossible. It's just annoying. It's even worse when you're in the subdirectory,
```
grep pattern *.d ../otherpackage/*.d ../whatever/*.d
```
yuck.
Besides, the modules in dmd are so interconnected I don't see much that'll genuinely split into packages. It's not even done right in the existing packages - `dmd.root.filename` is importing `dmd.utils`.

Reply:
>  It's just annoying

I still don't follow. Why can't you do sth. simple as:

```
cd dmd/src
grep -r pattern .
```

Of course, any decent editor would do that for you out of the box (and index all files for faster search queries) ...

> It's not even done right in the existing packages

That's not an argument against trying to improve the status quo.

Reply:
Let me put it another way. dmd is a tangle of code, with every module importing every other module. The existing package boundaries are not respected. I regularly fight off attempts to entangle the backend package with the front end. `dmd.root` importing code from `dmd` was written and approved without considering encapsulation principles. We as a team need to get a lot better.
Package-izing the front end without disentangling and doing proper encapsulation of dependencies is like beginning the restoration of an old car by painting over the rust, dirt, and dents. While it's gloriously fun to shoot color, it's a step backwards, not forwards, if the groundwork isn't done first.
Without doing the groundwork of encapsulation, you can't even determine what the packages should be.
Premature package-izing is an **Illusion of Progress**, involving massive disruption of the source code, build system, github history, rebasing, etc., and is likely to be completely wrong needing a do-over after the real grunt work is done.

Reply:
The modularity stuff is spot on. Sadly its purchase is diluted by the grep side story. Nobody should punish themselves and use `grep` instead of `git grep` in a git repo. (There should be a place in the story for `zsh`, too, with its nice `**` wildcards.)

Bottom line is, how about this simple plan - isolate first, repackage after. Sounds good?

Reply:
> Bottom line is, how about this simple plan - isolate first, repackage after. Sounds good?

That sounds nice in theory, but I don't know what is entangled other than a lot of it is. That still doesn't provide any clear benefit over the other way around, furthermore, packaging creates boundaries which guide the deconvolution of the code.

> Without doing the groundwork of encapsulation, you can't even determine what the packages should be.

What? I just listed a bunch of them: AST, sema, lib/scan (probably should go together?), glue, visitors, opts, iasm... There may be many more, but those are clear as day. The creation of one obviously-goes-together package is _completely orthogonal_ to the creation of _every other package_. If anything its the other way around, you can't disentangle something without knowing where you're going to place the pieces you're going to disentangle.

> and is likely to be completely wrong needing a do-over after the real grunt work is done.

Only if you do it wrong. Fortunately, as mentioned above, the creation of a package is orthogonal to the creation of every other package. As for the probability of it going wrong, you seriously overestimate it, it is lowered even further by the orthogonality of the problem.

Reply:
Packages also allow us to enforce e.g. `package(dmd.lib)` as a thing to force hard boundaries on code. This s particularly useful for reducing the cognitive burden for code that is modular-by-convention already in the same way the const-by-convention is a bad thing.

Reply:
Depending on how the packages are split it could help the Dub recipe file as well. Instead of explicitly listing files we could use wildcards to include all files in a package:

https://github.com/dlang/dmd/blob/95189b63de6644106fc564b747943a0999fd5f27/dub.sdl#L21-L31

Reply:
The way the dmd code is all entangled up, splitting modules into packages accomplishes nothing more than renaming `import statementsem;` to `import statement.sem;`. I don't see any evidence of anyone having the determination to follow through with enforcing package discipline - for example, `dmd.root.filename` is still importing `dmd.utils`. Without such followthru, packagizing dmd is a thorough waste of effort and massive disruption with zero benefit.

Reply:
I'd add: without automation, packaging is an illusion. We need scripts that prevent unwanted dependencies and run them automatically. Otherwise we do a major reshuffle followed by another one every couple of months.

No automatic enforcement means no package isolation.

Reply:
> No automatic enforcement means no package isolation.

We already have `package`, we just can't use it, because we aren't allowed to create packages :/

> We need scripts that prevent unwanted dependencies and run them automatically.

Well, we need the packages first, before we can have scripts that enforce their dependencies.

> Without such followthru, packagizing dmd is a thorough waste of effort and massive disruption with zero benefit.

Just as an example: We already have the `lexer` package which is separate from DMD (even built separately), so what's the point in not grouping these files under e.g. `dmd.lexer` and thus stronger enforcing that this separation stays?

Additionally, packages still help greatly to developers because they immediately know where to start looking. After all, we did split Phobos in subpackages for exactly that reason too...

Reply:
>We already have package, we just can't use it, because we aren't allowed to create packages :/

If you come with improved isolation by using package-level modules, that'd be great. I'm afraid it's more difficult than it seems. But if simple, a PR would be great to look at.

>Well, we need the packages first, before we can have scripts that enforce their dependencies.

We need infra in place and at the ready. Otherwise we repackage, nobody does the infra, and then packages start breaking their boundaries because there's no enforcement.

Reply:
> But if simple, a PR would be great to look at.

So that you can close it again like https://github.com/dlang/dmd/pull/7639?
No thank you very much.
The file that are used for the lexer package are: 

```
    "src/dmd/console.d" \
    "src/dmd/entity.d" \
    "src/dmd/errors.d" \
    "src/dmd/filecache.d" \
    "src/dmd/globals.d" \
    "src/dmd/id.d" \
    "src/dmd/identifier.d" \
    "src/dmd/lexer.d" \
    "src/dmd/tokens.d" \
    "src/dmd/utf.d"
```
`globals` and `errors` would probably go into a different package `utils` as its used not solely by the lexer.

> We need infra in place and at the ready. 

It's already there: we build _separate_ libraries (that can't depend on anything outside themselves). So if you were to call a function within the lexer that pulls in sth. else, the build will fail. Check this out: https://github.com/dlang/dmd/blob/master/dub.sdl

Additionally, for more fine-grained control we have `package(dmd.<package-name>):`. The process is:
- create package
- fix code to be in the package boundaries
- introduce automated checks

Not the other way. Also, we don't enforce package boundaries at Phobos and it's still very clear where a function is. There's no quick way to tell in which module  of DMD where the magic X happens.

Reply:
>Also, we don't enforce package boundaries at Phobos and it's still very clear where a function is. There's no quick way to tell in which module of DMD where the magic X happens.

Of course! The charters are very different. One is a library that's supposed to expose everything.

Reply:
No, it's because DMD's has no clear structures. Take @thewilsonator's comparison with clang (https://github.com/dlang/dmd/pull/9511#discussion_r270614236).
In clang you know where to look, in DMD you only have grep and luck. This is very frustrating and a huge waste of time. I will give you an example: where are struct literals processed in DMD?
In `expression.d` maybe? No.
In `expressionsem.d` maybe? No.
In `dstruct.d` maybe? No.
In `parse.d` maybe? No.
In `initsem.d`, obviously.
(there are many nastier examples)

Another benefit of packages in DMD is that e.g. the IR package could be excluded when building LDC or DMD as a library. Currently it's a nasty blacklist for dub.

Reply:
> If you come with improved isolation by using package-level modules, that'd be great.

Its called 

```
module dmd.foo.bar;
package(dmd.foo):
//rest of module
````

> I'm afraid it's more difficult than it seems. But if simple, a PR would be great to look at.

I don't have time ATM but `dmd/lib.d` & friends should already be separated.

Reply:
>The way the dmd code is all entangled up, splitting modules into packages accomplishes nothing more than renaming import statementsem; to import statement.sem;. 

First of all it'd be `import dmd.sem.statement`, packaging with that heirarchy would be lunacy...

>I don't see any evidence of anyone having the determination to follow through with enforcing package discipline - for example, dmd.root.filename is still importing dmd.utils. Without such followthru, packagizing dmd is a thorough waste of effort and massive disruption with zero benefit.

... and secondly, that is patently false: it makes it vastly easier to navigate, especially for newcomers. I was talking to someone the other day who occasionally runs into complier bugs, and is more that competent enough to fix them, and asked why he doesn't. His reply was that the DMD codebase is, on the package level, completely impenetrable and he does not feel that delving through the ~120 files is a good use of his time. 

Let me state that again differently so that it sinks in: **DMD's package malorganisation is actively stopping potential contributors from contributing**.

Enforcing package boundaries in nice, and something we should be doing, but rejecting the creation of packages in DMD is like rejecting the first medicine in a course of treatment because you haven't had the second one yet. You can't enforce package boundaries if you don't have any, conversely having packages makes it easier to enforce the boundaries between them.

Reply:
[This](https://github.com/lempiji/ddeps) was Just announced, should help us assess the situation.

Reply:
What if you split the modules into packages, and created a flat directory of symlinks for Walter? i.e.

```
src/dmd/pkg/visitor/transitive.d
src/dmd/pkg/visitor/permissive.d
src/dmd/pkg/visitor/strict.d

src/dmd/transitivevisitor.d -> pkg/visitor/transitive.d
src/dmd/permissivevisitor.d -> pkg/visitor/permissive.d
src/dmd/strictvisitor.d -> pkg/visitor/strict.d
```

Reply:
Symlinks don't work that well on Windows?

Reply:
I don't think Walter uses windows.

Reply:
I assume Windows is his main platform. You do know that DMD existed first on Windows? And that DMC only exists on Windows?

Reply:
Oh...I guess I thought it was originally written for Linux since it's do much easier to develop dmd on Linux.

Reply:
No. It's just that he has his own way of developing on Windows compared to how everyone else is doing Windows development. Custom compiler, linker, object format and so on. But on Linux he kind of follows what most developers are doing.

Reply:
> Symlinks don't work that well on Windows?

They don't? Wikipedia seems to imply there aren't any major issues. https://en.wikipedia.org/wiki/NTFS_symbolic_link

Reply:
> "The default security settings in Windows Vista/Windows 7 disallow non-elevated administrators and all non-administrators from creating symbolic links. [..]
Starting with Windows 10 Insiders build 14972 the requirement for elevated administrator privileges was removed, allowing symlinks to be created without needing to elevate the console as administrator [1].

[..]
Windows Installer does not fully support symbolic links. A redirected \Windows\Installer will cause most .msi-based Windows installers to fail with error 2755 and/or error 1632."

In general symlinks / junction points are a can of worms on Windows. Mostly because people expect them to work just like on Posix systems, but that's far from the case.

Just as single data point, Git for Windows' support symlinks is disabled by default.

Reply:
If we're serious about this (I was only half serious) then it only needs to work on Walter's box :)